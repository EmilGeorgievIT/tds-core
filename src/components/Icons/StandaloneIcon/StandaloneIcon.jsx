import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../../utils/safeRest'

import Icon from '../Icon/Icon'

import styles from './StandaloneIcon.modules.scss'

/**
 * An icon that has meaning within the context of the page, which should be communicated to screen readers.
 *
 * <span class="docs--badge__new">new!</span> <span class="docs--badge__version">v0.24.0</span>
 */
class StandaloneIcon extends React.Component {
  onClick = event => {
    const {onClick} = this.props

    if (onClick) {
      onClick(event)
    }
  }

  render() {
    const {symbol, variant, size, interactive, onClick, a11yText, ...rest} = this.props

    if (interactive) {
      return (
        <button {...safeRest(rest)} className={styles.interactive} aria-label={a11yText}>
          <Icon symbol={symbol} variant={variant} size={size} />
        </button>
      )
    }

    return <Icon {...rest} symbol={symbol} variant={variant} size={size} aria-label={a11yText} />
  }
}

StandaloneIcon.propTypes = {
  /**
   * Name of the icon symbol.
   */
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times',
  ]).isRequired,
  /**
   * The appearance of the Icon.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted', 'error']),
  /**
   * The icon size in pixels.
   */
  size: PropTypes.oneOf([16, 24, 48]),
  /**
   * Make the icon interactive
   */
  interactive: PropTypes.bool,
  /**
  * A callback function to be invoked when the interactive icon is clicked.
  */
  onClick: PropTypes.func,
  /**
   * A description of the icon for screen readers.
   */
  a11yText: PropTypes.string.isRequired,
}

StandaloneIcon.defaultProps = {
  variant: undefined,
  size: 24,
  interactive: false,
  onClick: undefined,
}

export default StandaloneIcon
