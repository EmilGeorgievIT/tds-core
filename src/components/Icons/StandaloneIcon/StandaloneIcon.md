### Available Icons

```
<div className="docs--horizontal-spacing">
  <StandaloneIcon symbol="caretDown" a11yText="text" />
  <StandaloneIcon symbol="caretUp" a11yText="text" />
  <StandaloneIcon symbol="checkmark" a11yText="text" />
  <StandaloneIcon symbol="leftChevron" a11yText="text" />
  <StandaloneIcon symbol="chevron" a11yText="text" />
  <StandaloneIcon symbol="exclamationPointCircle" a11yText="text" />
  <StandaloneIcon symbol="expander" a11yText="text" />
  <StandaloneIcon symbol="hamburger" a11yText="text" />
  <StandaloneIcon symbol="incomplete" a11yText="text" />
  <StandaloneIcon symbol="location" a11yText="text" />
  <StandaloneIcon symbol="minus" a11yText="text" />
  <StandaloneIcon symbol="plus" a11yText="text" />
  <StandaloneIcon symbol="questionMarkCircle" a11yText="text" />
  <StandaloneIcon symbol="spyglass" a11yText="text" />
  <StandaloneIcon symbol="times" a11yText="text" />
</div>
```
