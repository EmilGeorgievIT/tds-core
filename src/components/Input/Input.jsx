import React from 'react'
import PropTypes from 'prop-types'
import safeRest from '../../safeRest'

import styles from './Input.modules.scss'

const textToId = text => text.toLowerCase().replace(/ /g, '-')

class Input extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      value: this.props.value
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.state.value !== nextProps.value) {
      this.setState({
        value: nextProps.value
      })
    }
  }

  onChange = (event) => {
    const { onChange } = this.props

    this.setState({
      value: event.target.value
    })

    if (onChange) {
      onChange(event)
    }
  }

  render() {
    const { type, label, ...rest } = this.props

    const id = rest.id || rest.name || textToId(label)

    return (
      <div>
        <label htmlFor={id} className={styles.label}>{label}</label>
        <input
          {...safeRest(rest)} id={id} type={type} className={styles.input}
          value={this.state.value} onChange={this.onChange}
        />
      </div>
    )
  }
}

Input.propTypes = {
  type: PropTypes.oneOf(['text', 'number', 'password']), // TODO: finish this list
  label: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  onChange: PropTypes.func
}

Input.defaultProps = {
  type: 'text',
  value: undefined,
  onChange: undefined
}

export default Input
