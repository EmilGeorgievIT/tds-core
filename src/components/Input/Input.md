```
<div>
  <Input label="First name (default)" />

  <Input type="password" label="Password" value="123abc" />
  <Input type="number" label="Age" value="32" />

  <Input label="First name (auto focus)" autoFocus />

  <Input label="First name (disabled)" disabled />

  <Input label="First name (success)" feedback="success" />
  <Input label="First name (error)" feedback="error" />

  <Input label="First name (error msg)" feedback="error" error="First name is required" />

  <Input label="First name (helper)" feedback="success"
    helper={<Input.Helper>Some helper stuff</Input.Helper>}
  />
</div>
```
