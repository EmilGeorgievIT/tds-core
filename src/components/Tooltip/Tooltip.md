```
<Input
  label="Full name"
  tooltip={
    <Tooltip>Enter your first name and last name.</Tooltip>
  }
/>
```

```
<Input
  label="Full name 2"
  tooltip={
    <Tooltip direction="left">Enter your first name and last name.</Tooltip>
  }
/>
```
