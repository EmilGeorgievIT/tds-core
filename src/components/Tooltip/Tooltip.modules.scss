@import '../../scss/settings/colours';

$bubble-trigger-size: 16px;

$bubble-triangle-height: 10px;
$bubble-triangle-width: 7px;
$bubble-triangle-position: $bubble-triangle-width + 5px;

$bubble-position: $bubble-trigger-size * -0.5;

.wrapper {
  composes: fixLineHeight from '../Icons/IconWrapper.modules.scss';

  position: relative;
}

.trigger {
  composes: none from '../Borders.modules.scss';

  background: none;
  box-shadow: none;
  cursor: pointer;

  // FIXME: Without the 1px padding the icon gets clipped just barely. This is due to the odd sizing/positioning of
  //  the icon font
  // composes: noSpacing from '../Spacing.modules.scss';
  padding: 0 1px;
}

.baseBubble {
  composes: rounded from '../Borders.modules.scss';

  position: absolute;
  bottom: calc(100% + #{$bubble-triangle-height});

  min-width: 200px;
  max-width: 600px; // TODO: check what the value should be

  background-color: $color-white;
  box-shadow: 0 0 2px 0 $color-shark, 0 3px 2px 0 rgba(84, 89, 95, 0.25);

  // This is the bubble triangle
  &::before {
    display: block;
    position: absolute;
    content: '';

    bottom: $bubble-triangle-width * -1;

    border-width: $bubble-triangle-width;
    border-style: solid;
    border-color: transparent $color-white $color-white transparent;

    background-color: $color-white;
    box-shadow: 2px 2px 3px 0 rgba(42, 42, 44, 0.4);

    transform: rotate(45deg);
  }
}

.left {
  composes: baseBubble;

  right: $bubble-position;

  &::before {
    right: $bubble-triangle-position;
  }
}

.right {
  composes: baseBubble;

  left: $bubble-position;

  &::before {
    left: $bubble-triangle-position;
  }
}
