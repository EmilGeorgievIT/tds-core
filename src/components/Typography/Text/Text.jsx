import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../../utils/safeRest'
import TextSup from './TextSup/TextSup'
import TextSub from './TextSub/TextSub'

import styles from './Text.modules.scss'

/**
 * Produces inline text
 *
 * <span class="docs--badge green">new!</span> <span class="docs--badge purple">v0.22.0</span>
 */
const Text = ({ bold, size, invert, children, ...rest }, context) => {
  const textColor = invert ? styles.invertedColor : styles.color

  const classes = `
    ${styles[size]}
    ${bold ? styles.boldFont : styles[`${size}Font`]}
    ${context.inheritColor ? styles.inheritColor : textColor}
  `

  return (
    <span {...safeRest(rest)} className={classes}>
      {children}
    </span>
  )
}

Text.propTypes = {
  /**
   * Embolden text without conveying any special importance or relevance.
   */
  bold: PropTypes.bool,
  /**
   * Font size.
   */
  size: PropTypes.oneOf([
    'base',
    'small',
    'medium',
    'large'
  ]),
  /**
   * Invert span style to appear light on dark backgrounds.
   */
  invert: PropTypes.bool,
  /**
   * The text.
   */
  children: PropTypes.node.isRequired
}

Text.defaultProps = {
  bold: false,
  size: 'base',
  invert: false
}

Text.contextTypes = {
  inheritColor: PropTypes.bool
}

Text.Sup = TextSup
Text.Sub = TextSub

export default Text
