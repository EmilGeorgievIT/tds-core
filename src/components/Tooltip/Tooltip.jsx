import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../utils/safeRest'
import joinClassNames from '../../utils/joinClassNames'
import generateId from '../../utils/generateId'
import closest from './element-closest'

import StandaloneIcon from '../Icons/StandaloneIcon/StandaloneIcon'
import Text from '../Typography/Text/Text'
import Box from '../Box/Box'

import displayStyles from '../Display.modules.scss'
import iconWrapperStyles from '../Icons/IconWrapper.modules.scss'
import styles from './Tooltip.modules.scss'

const getTriggerA11yText = connectedFieldLabel => {
  if (!connectedFieldLabel) {
    return 'Reveal additional information.'
  }

  return `Reveal additional information about ${connectedFieldLabel}.`
}

const getIds = connectedFieldLabel => {
  const id = generateId(connectedFieldLabel, 'unknown-field')

  return {
    bubbleId: id.postfix('tooltip'),
    triggerId: id.postfix('trigger'),
  }
}

/**
 * Provide an explanation or instructions for a form field that most users do not need.
 *
 * <span class="docs--badge__new">new!</span> <span class="docs--badge__version">v0.24.0</span>
 */
class Tooltip extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      open: false,
    }
  }

  componentDidUpdate() {
    if (this.state.open) {
      document.addEventListener('click', this.toggleBubbleOnOutsideEvent)
      document.addEventListener('keypress', this.toggleBubbleOnOutsideEvent)
    } else {
      document.removeEventListener('click', this.toggleBubbleOnOutsideEvent)
      document.removeEventListener('keypress', this.toggleBubbleOnOutsideEvent)
    }
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.toggleBubbleOnOutsideEvent)
    document.removeEventListener('keypress', this.toggleBubbleOnOutsideEvent)
  }

  toggleBubbleOnOutsideEvent = event => {
    const { connectedFieldLabel } = this.props

    const { bubbleId, triggerId } = getIds(connectedFieldLabel)

    const inBubble = closest(event.target, `#${bubbleId}`)
    const inTrigger = closest(event.target, `#${triggerId}`)

    if (!inBubble && !inTrigger) {
      this.toggleBubble()
    }
  }

  toggleBubble = () => {
    this.setState(({ open }) => {
      return { open: !open }
    })
  }

  renderBubble(id, direction, open, content) {
    const classes = joinClassNames(styles[direction], !open && displayStyles.hide)

    return (
      <Box
        spacing="padding"
        vertical={2}
        horizontal={3}
        dangerouslyAddClassName={classes}
        id={id}
        role="tooltip"
        aria-live="polite"
        aria-hidden={open ? 'false' : 'true'}
        data-testid="bubble"
      >
        <Text size="small">{content}</Text>
      </Box>
    )
  }

  render() {
    const { direction, connectedFieldLabel, children, ...rest } = this.props

    const { bubbleId, triggerId } = getIds(connectedFieldLabel)

    return (
      <div {...safeRest(rest)} className={iconWrapperStyles.fixLineHeight}>
        {this.renderBubble(bubbleId, direction, this.state.open, children)}

        <button
          id={triggerId}
          className={styles.trigger}
          onClick={this.toggleBubble}
          aria-controls={bubbleId}
          aria-haspopup="true"
          aria-expanded={this.state.open ? 'true' : 'false'}
        >
          <StandaloneIcon
            symbol="questionMarkCircle"
            a11yText={getTriggerA11yText(connectedFieldLabel)}
          />
        </button>
      </div>
    )
  }
}

Tooltip.propTypes = {
  /**
   * Open the bubble to the left or right of the trigger.
   */
  direction: PropTypes.oneOf(['left', 'right']),
  /**
   * The input field that using this tooltip must pass its label so that the tooltip trigger can connect itself. Do not
   * show this prop as it is not part of the public API. If this prop is not specified, the tooltip will generate
   * a generic ID, which could produce duplicate ids.
   *
   * @ignore
   */
  connectedFieldLabel: PropTypes.string,
  /**
   * The message. Can be raw text or text components.
   */
  children: PropTypes.node.isRequired,
}

Tooltip.defaultProps = {
  direction: 'right',
  connectedFieldLabel: undefined,
}

export default Tooltip
