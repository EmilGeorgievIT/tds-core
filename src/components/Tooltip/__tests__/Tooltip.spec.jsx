import React from 'react'
import { shallow, render, mount } from 'enzyme'


import StandaloneIcon from '../../Icons/StandaloneIcon/StandaloneIcon'
import Text from '../../Typography/Text/Text'

import Input from '../../Input/Input'
import Tooltip from '../Tooltip'

describe('Tooltip', () => {
  const defaultChildren = 'Tooltip text'
  const doShallow = (overrides = {}, children = defaultChildren) =>
    shallow(<Tooltip {...overrides}>{children}</Tooltip>)

  const findBubble = tooltip => tooltip.find('[data-testid="bubble"]')
  const findTrigger = tooltip => tooltip.find(StandaloneIcon)
  const toggleBubble = tooltip => findTrigger(tooltip).simulate('click')

  it('renders', () => {
    const tooltip = render(<Tooltip>Helper text</Tooltip>)

    expect(tooltip).toMatchSnapshot()
  })

  it('has a trigger', () => {
    const tooltip = doShallow()

    expect(findTrigger(tooltip)).toHaveProp('symbol', 'questionMarkCircle')
    expect(findTrigger(tooltip)).toHaveProp('a11yText', 'Reveal additional information.')
  })

  it('has small text in the bubble', () => {
    const tooltip = doShallow({}, 'Some content')
    toggleBubble(tooltip)

    expect(findBubble(tooltip)).toContainReact(<Text size="small">Some content</Text>)
  })

  it('has a direction', () => {
    let tooltip = doShallow()
    expect(findBubble(tooltip).dive()).toHaveClassName('right')

    tooltip = doShallow({ direction: 'left' })
    expect(findBubble(tooltip).dive()).toHaveClassName('left')
  })

  describe('interactivity', () => {
    const createRootElement = () => {
      const root = document.createElement('div')
      root.setAttribute('id', `tooltip-${Math.random()}`)

      return root
    }

    const getBubbleClassNames = tooltip =>
      findBubble(tooltip)
        .at(1) // Not sure why it finds two...
        .render()
        .attr('class')

    let mountedToolTip

    afterEach(() => {
      if (mountedToolTip) {
        mountedToolTip.unmount()
      }
    })

    it('shows and hides the bubble', () => {
      const tooltip = doShallow()
      expect(findBubble(tooltip).dive()).toHaveClassName('hide')

      toggleBubble(tooltip)
      expect(findBubble(tooltip).dive()).not.toHaveClassName('hide')

      toggleBubble(tooltip)
      expect(findBubble(tooltip).dive()).toHaveClassName('hide')
    })

    it('hides the bubble when clicking outside of it', () => {
      const root = createRootElement()

      mountedToolTip = mount(<Tooltip>Tooltip text</Tooltip>, {
        attachTo: document.body.appendChild(root),
      })

      toggleBubble(mountedToolTip)
      root.dispatchEvent(new Event('click', { bubbles: true }))

      expect(getBubbleClassNames(mountedToolTip)).toContain('hide')

      // And it removes the event listener so on subsequent clicks it remains hidden
      root.dispatchEvent(new Event('click'), { bubbles: true })

      expect(getBubbleClassNames(mountedToolTip)).toContain('hide')
    })

    it('will not hide the bubble when clicking inside the bubble', () => {
      const root = createRootElement()

      mountedToolTip = mount(<Tooltip>Tooltip text</Tooltip>, {
        attachTo: document.body.appendChild(root),
      })

      toggleBubble(mountedToolTip)
      root
        .querySelector('[data-testid="bubble"]')
        .dispatchEvent(new Event('click', { bubbles: true }))

      expect(getBubbleClassNames(mountedToolTip)).not.toContain('hide')
    })
  })

  describe('accessibility', () => {
    it('connects the bubble message to the trigger button for screen readers', () => {
      const tooltip = doShallow({ connectedFieldLabel: 'Some field' })

      expect(findBubble(tooltip)).toHaveProp('id', 'some-field_tooltip')
      expect(findBubble(tooltip)).toHaveProp('role', 'tooltip')
      expect(findBubble(tooltip)).toHaveProp('aria-live', 'polite')

      expect(findTrigger(tooltip)).toHaveProp('aria-haspopup', 'true')
      expect(findTrigger(tooltip)).toHaveProp('aria-controls', 'some-field_tooltip')
    })

    it('shows and hides the bubble for screen readers', () => {
      const tooltip = doShallow()

      expect(findTrigger(tooltip)).toHaveProp('aria-expanded', 'false')
      expect(findBubble(tooltip)).toHaveProp('aria-hidden', 'true')

      toggleBubble(tooltip)

      expect(findTrigger(tooltip)).toHaveProp('aria-expanded', 'true')
      expect(findBubble(tooltip)).toHaveProp('aria-hidden', 'false')
    })
  })

  it('passes additional attributes to the element', () => {
    const tooltip = doShallow({ 'data-some-attr': 'some value' })

    expect(tooltip).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const tooltip = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(tooltip).not.toHaveProp('className', 'my-custom-class')
    expect(tooltip).not.toHaveProp('style')
  })
})

describe('Connecting Tooltips to form fields', () => {
  it('has a default when not connected to any form field', () => {
    const tooltip = shallow(<Tooltip>The tooltip</Tooltip>)

    expect(tooltip.find(StandaloneIcon)).toHaveProp('a11yText', 'Reveal additional information.')
  })

  it('connects to Input', () => {
    const input = shallow(<Input label="Some field" tooltip={<Tooltip>The tooltip</Tooltip>} />)

    expect(
      input
        .find(Tooltip)
        .dive()
        .find(StandaloneIcon)
    ).toHaveProp('a11yText', 'Reveal additional information about Some field.')
  })
})
