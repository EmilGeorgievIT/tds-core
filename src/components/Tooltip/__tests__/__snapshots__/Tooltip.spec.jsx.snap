// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
<div
  class="fixLineHeight"
  id="the-id"
>
  <div
    aria-hidden="true"
    aria-live="polite"
    class="verticalPadding-2 horizontalPadding-3 right hide"
    data-testid="bubble"
    id="unknown-field_tooltip"
    role="tooltip"
  >
    <span
      class="small smallFont color"
    >
      Helper text
    </span>
  </div>
  <button
    aria-controls="unknown-field_tooltip"
    aria-expanded="false"
    aria-haspopup="true"
    class="trigger"
  >
    <i
      aria-label="Reveal additional information."
      class="iconQuestionMarkCircle size24"
    >
      
    </i>
  </button>
</div>
`;
